fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/SizeStatisticsCollector.java

```diff
@@ -93,28 +80,15 @@ private SizeStatisticsCollector(@Nonnull SubspaceProvider subspaceProvider) {
     @Nonnull
     public CompletableFuture<Boolean> collectAsync(@Nonnull FDBRecordContext context, @Nonnull ExecuteProperties executeProperties) {
         if (continuation.isEnd()) {
-            return AsyncUtil.READY_TRUE;
+            return CompletableFuture.completedFuture(true);
         }
         return subspaceProvider.getSubspaceAsync(context).thenCompose(subspace -> {
             final ScanProperties scanProperties = new ScanProperties(executeProperties)
                     .setStreamingMode(CursorStreamingMode.WANT_ALL);
-            final KeyValueCursor kvCursor = KeyValueCursor.Builder.withSubspace(subspace)
-                    .setContext(context)
-                    .setContinuation(continuation.toBytes())
-                    .setScanProperties(scanProperties)
-                    .build();
+            final SizeStatisticsCollectorCursor statsCursor = SizeStatisticsCollectorCursor.ofSubspace(subspace, context, scanProperties, continuation.toBytes());
 
-            return kvCursor.forEachResult(nextResult -> {
-                KeyValue kv = nextResult.get();
-                keyCount += 1;
-                keySize += kv.getKey().length;
-                maxKeySize = Math.max(maxKeySize, kv.getKey().length);
-                valueSize += kv.getValue().length;
-                maxValueSize = Math.max(maxValueSize, kv.getValue().length);
-                int totalSize = kv.getKey().length + kv.getValue().length;
-                if (totalSize > 0) {
-                    sizeBuckets[Integer.SIZE - Integer.numberOfLeadingZeros(totalSize) - 1] += 1;
-                }
+            return statsCursor.forEachResult(nextResult -> {
+                sizeStatsResults = nextResult.get();
```

Comment:
We only get a nextResult with a value is when we have fully finished aggregation. Otherwise, we fall into the handle() method and return only a continuation.

Reply:
If the cursor emitted partial results instead of no value and a continuation at each onNext() iteration then would need to aggregate the partial results.

Reply:
Yeah, sorry, I think I left this comment when I was still working through how everything worked, and this looked suspect given how I assumed it had been implemented. But I think the way it is actually implemented, this makes sense.